include/lbann/callbacks/summarize_images.hpp

```diff
@@ -0,0 +1,221 @@
+ *  @brief Dump images to event files based on strategy
+ */
+class autoencoder_strategy : public image_output_strategy {
+
+public:
+
+  /** @brief autoencoder_strategy : image_output_strategy Constructor.
+   *  @param sample_indices Vector of sample indices for images
+   */
+  autoencoder_strategy(std::string const& input_layer_name,
+                       size_t num_images = 10)
+    : m_input_layer_name{input_layer_name},
+      m_num_images{num_images} {}
+
+  /** @brief Get vector containing indices of images to be dumped.
+   *  @returns std::vector<int> Vector with indices of images to dump.
+   */
+  std::vector<std::pair<size_t, El::Int>>
+  get_image_indices(model const& m) const final;
+
+  /** @brief Construct tag for image */
+  std::string get_tag(std::string const& layer_name,
+                      El::Int index, El::Int epoch) const final;
+
+private:
+
+  /** @brief Name of input layer */
+  std::string m_input_layer_name;
+
+  /** @brief Number of images to be tracked */
+  size_t m_num_images;
+
+  /** @brief Sample indices of images to track */
+  mutable std::unordered_set<El::Int> m_tracked_images;
+
+  /** @brief A map from models to shuffled indices */
+  mutable std::unordered_map<model const*, std::vector<size_t>> m_shuffled_indices;
+
+}; // class Autoencoder : image_output_strategy
+
+std::unique_ptr<image_output_strategy>
+build_autoencoder_strategy_from_pbuf(google::protobuf::Message const&);
+
+/** @class summarize_images
+ *  @brief Callback to dump images to event files based on strategy
+ */
+class summarize_images : public callback_base {
+
+public:
+  /** @brief summarize_images Constructor.
+   *  @param summarizer Pointer to lbann_summary object
+   *  @param strategy Pointer to image image_output_strategy
+   *  @param img_source_layer_name Name of image layer
+   *  @param interval Interval of epochs to dump images
+   *  @param img_format Image file format (e.g. .jpg, .png, .pgm)
+   */
+  summarize_images(std::shared_ptr<lbann_summary> const& summarizer,
+                   std::shared_ptr<image_output_strategy> const& strategy,
+                   std::string const& img_source_layer_name,
+                   uint64_t interval = 1,
+                   std::string const& img_format = ".jpg");
+
+  /** @brief Copy constructor */
+  callback_base* copy() const override { return new summarize_images(*this); }
+
+  /** @brief Return name of callback */
+  std::string name() const override { return "summarize_images"; }
+
+  /** @brief Hook to pull data from lbann run */
+  void on_batch_evaluate_end(model* m) override;
+
+private:
+
+  /** @brief Add image to event file */
+  void dump_images_to_summary(model const& m) const;
+
+
+private:
+
+  /* @brief lbann_summary object */
+  std::shared_ptr<lbann_summary> m_summarizer;
+
+  /* @brief image_output_strategy object */
+  std::shared_ptr<image_output_strategy> m_strategy;
```

Comment:
Is this actually shared?
```suggestion
  std::unique_ptr<image_output_strategy> m_strategy;
```